### ISSUE TITLE
How to pack models generated during bagging together?

### ISSUE THREAD

**Issue by KleinXin** [awslabs/autogluon#725]
By default, num_bagging_folds is 10 and it will generate 10 models. A model named model.pkl will use all sub-models in each folder during inference. An example is shown below.

![image](https://user-images.githubusercontent.com/13084622/96829565-84162780-146c-11eb-9352-dffe7e673db8.png)

The problem is that the path of all sub-models can not be changed when model.pkl uses them. For example, when all models are saved in folder './res' the suffix of the path of each sub-model is something like './res/models/CatboostClassifier_STACKER_l1/CatboostClassifier_fold_x'. This is not so convenient to integrate the model into the system.

Is there any way to pack all these sub-models together?

To solve this problem, now I set auto_stack to False so that only one model is generated. This model can be saved in any folder. However, the auc is a little bit lower than ensembled model.

Hope it has a better solution to this problem. thx!

**Comment by Innixma** [comment]
Could you explain why you need to access the individual models? What is the problem you are having?

You can do:

predictor = task.fit(...)

`predictor.predict(test_data)`

or

`predictor.predict(test_data, model='CatboostClassifier_STACKER_l1')`

Alternatively, you can compress the bags into a single model via `predictor.refit_full()`, but with a slight reduction in accuracy.

**Comment by KleinXin** [comment]
> Could you explain why you need to access the individual models? What is the problem you are having?
> 
> You can do:
> 
> ```
> predictor = task.fit(...)
> 
> ensemble_preds = predictor.predict(test_data)
> 
> bagged_catboost_preds = predictor.predict(test_data, model='CatboostClassifier_STACKER_l1')
> ```
> 
> Alternatively, you can compress the bags into a single model via `predictor.refit_full()`, but with a slight reduction in accuracy.
> 
> If you want to move the files, just move all of the files in the model directory together. The loading of the fold models is based on relative path.

My Problem is that the paths of all the sub-models are not relative paths. For example, suppose the path of the model.pkl in the figure above is './res/20201022/model.pkl'. Then I copy model.pkl to './res' and load './res/model.pkl', the name of folder 20201022 is changed to 20201023 at the same time. So in the end the path of the model in folder 'CatboostClassifier_fold_0' is './res/20201023/CatboostClassifier_fold_0/model.pkl'.

When I load './res/model.pkl', it will give an error which says that './res/20201022/CatboostClassifier_fold_0/model.pkl' is missing. That is why I want to pack all models together so that I can put the model anywhere without those models in folders CatboostClassifier_fold_x.

Anyway, I will try to use predictor.refit_full() as you suggested.

**Comment by Innixma** [comment]
Why are you loading model.pkl? It is **highly recommended** to just use the predictor object, as numerous things can go wrong by using the model files in isolation (Without predictor, the model will not replicate the feature engineering done during training). So long as you move the entire folder directory of the predictor it will work.

Regarding your example, this fundamentally cannot work regardless of how it is implemented. It is impossible for './res/model.pkl' to know that you changed the directory name to '/20201023/' for the fold models.

**Comment by KleinXin** [comment]
> Why are you loading model.pkl? It is **highly recommended** to just use the predictor object, as numerous things can and will go wrong by using the model files in isolation (Without predictor, the model will not replicate the feature engineering done during training). So long as you move the entire folder directory of the predictor it will work.
> 
> Regarding your example, this fundamentally cannot work regardless of how it is implemented. It is impossible for './res/model.pkl' to know that you changed the directory name to '/20201023/' for the fold models.

In fact, we want to use the model trained by AutoGluon but donot want to integrate AutoGluon into our system. So I used pickle to load the model file. 

I think it is better to give the hyperparameters about how to train the model just like tpot does. However, AutoGluon may not support that because bagging is also used in trainning process. Is that right?

**Comment by Innixma** [comment]
What is the restriction in terms of not integrating AutoGluon? If it is due to dependencies this has been made easier in upcoming v0.1 through modularization. Trying to extract the model from AutoGluon is a complex process and will only work if you also include the feature generator. The model file is dependent on AutoGluon as it is an AutoGluon AbstractModel implementation.

To get the model hyperparameters call predictor.leaderboard(extra_info=True) or predictor.info(). Hyperparameters are the same for all bagged models except for iterations, which are averaged in the results. You can also get per-fold hyperparameters with predictor.info().

**Comment by KleinXin** [comment]
@Innixma Thank you for your reply!
Our system is using GO with XGBoost. So we have to retrain the model using XGBoost or CatBoost with parameters searched by AutoGluon so that it can be loaded.

I tried to use predictor.leaderboard(extra_info=True). I think all parameters regarding trainning can be acquired from key 'model_info' of predictor.info() dict. 

I will print them out and study how to train the model using those parameters.

**Comment by Innixma** [comment]
Ah, I see. In that case predictor.info() will be your best resource and should hopefully allow you to replicate the majority of results. Just keep in mind that AutoGluon does perform feature engineering/preprocessing that you should analyze in order to understand what is being fed into the model. Given your restrictions, I also recommend that you avoid using stack ensembles as they will be **_very_** complex to implement outside of AutoGluon, and instead leverage HPO via hyperparameter_tune=True

**Comment by KleinXin** [comment]
> Ah, I see. In that case predictor.info() will be your best resource and should hopefully allow you to replicate the majority of results. Just keep in mind that AutoGluon does perform feature engineering/preprocessing that you should analyze in order to understand what is being fed into the model. Given your restrictions, I also recommend that you avoid using stack ensembles as they will be **_very_** complex to implement outside of AutoGluon, and instead leverage HPO via hyperparameter_tune=True
> 
> If you are interested in XGBoost, we are adding XGBoost into AutoGluon soon: #691

Thank you for your suggestion! 
I also realize that ensemble is very complex. So I set num_bagging_folds to 0 and auto_stack to False so that only one model is generated.
